x/cosmwasmpool/README.md

```diff
@@ -0,0 +1,284 @@
+/// The amount of tokens out is determined by the current exchange rate and the swap fee.
+/// The user specifies a minimum amount of tokens out, and the transaction will revert if that amount of tokens
+/// is not received.
+SwapExactAmountIn {
+    sender: String,
+    token_in: Coin,
+    token_out_denom: String,
+    token_out_min_amount: Uint128,
+    swap_fee: Decimal,
+},
+/// SwapExactAmountOut swaps as many tokens in as possible for an exact amount of tokens out.
+/// The amount of tokens in is determined by the current exchange rate and the swap fee.
+/// The user specifies a maximum amount of tokens in, and the transaction will revert if that amount of tokens
+/// is exceeded.
+SwapExactAmountOut {
+    sender: String,
+    token_in_denom: String,
+    token_in_max_amount: Uint128,
+    token_out: Coin,
+    swap_fee: Decimal,
+},
+```
+
+The reason why this needs to be sudo endpoint, which can only be called by the chain itself, is that the chain can provide correct information about `swap_fee`, which can be deviated from contract defined `swap_fee` in multihop scenario.
+
+(TBD)
+Problem with this approach is that swap expects funds to be sent with the same amount as `token_in` but there is no way to validate that in contract, so we need to make sure of that in `x/cosmwasmpool` module.
+
+Alternative approach is to expose swap in execute endpoint instead and validate that msg sender is a specific module account so that we can leverage `MsgExecuteContract`'s `funds` field to send funds to the contract.
+
+## Deactivating
+
+On contract's sudo enpoint, `SetActive` can be called to deactivate the pool. This will prevent the pool from being used for swap, and also prevent users from providing liquidity to the pool. Contract needs to check if the pool is active before performing any state mutating operation except `SetActive`.
+
+```rs
+SetActive {
+    is_active: bool,
+}
+```
+
+(TBD) On how to handle the deactivation operationally.
+
+## CosmWasm Pool Contract Interface
+
+The contract interface is defined so that `cosmwasmpool` can delegate `PoolI` and `PoolModuleI` calls to contract.
+
+The following are the messages that the contract needs to implement. (If you have trouble interpreting this, please read [Rust de/serialization](#rust-deserialization))
+
+### Query
+```rs
+#[cw_serde]
+#[derive(QueryResponses)]
+enum QueryMessage {
+    /// GetSwapFee returns the pool's swap fee, based on the current state.
+    /// Pools may choose to make their swap fees dependent upon state
+    /// (prior TWAPs, network downtime, other pool states, etc.)
+    /// hence Context is provided as an argument.
+    #[returns(GetSwapFeeResponse)]
+    GetSwapFee {},
+
+    /// Returns whether the pool has swaps enabled at the moment
+    #[returns(IsActiveResponse)]
+    IsActive {},
+
+    /// GetTotalShares returns the total number of LP shares in the pool
+    #[returns(TotalSharesResponse)]
+    GetTotalShares {},
```

Comment:
This is not on the `PoolI` anymore. Ref: https://github.com/osmosis-labs/osmosis/pull/4801

Reply:
`PoolI` is a pool interface that is implemented by pool models. These pool models are also written to state.

Examples of pool models:
- balancer
- stableswap
- concentrated liquidity

we are going to add a cosmwasm pool model.

`PoolModuleI` is an interface implemented by the SDK module keepers encapsulating these pool models.

We have:
- x/gamm keeper
- x/concentrated-liquidity keeper

we are going to add:
- x/cosmwasm keeper